scanners/sslyze.py

```diff
@@ -224,80 +223,83 @@ def parse_sslyze(raw_json):
-        elif isinstance(leaf_key, ec.EllipticCurvePublicKey):
-            leaf_key_type = "ECDSA"
+    if (constructed_chain and (len(constructed_chain) > 0)):
+        highest_constructed = parse_cert(constructed_chain[-1])
+        issuer = cert_issuer_name(highest_constructed)
+        if issuer:
+            data['certs']['constructed_issuer'] = issuer
         else:
-            leaf_key_type == str(leaf_key.__class__)
+            data['certs']['constructed_issuer'] = "(None constructed)"
 
-        data['certs']['key_type'] = leaf_key_type
+    leaf = parse_cert(served_chain[0])
+    leaf_key = leaf.public_key()
 
-        # Signature of the leaf certificate only.
-        data['certs']['leaf_signature'] = leaf.signature_hash_algorithm.name
+    if hasattr(leaf_key, "key_size"):
+        data['certs']['key_length'] = leaf_key.key_size
+    elif hasattr(leaf_key, "curve"):
+        data['certs']['key_length'] = leaf_key.curve.key_size
+    else:
+        data['certs']['key_length'] = None
+
+    if isinstance(leaf_key, rsa.RSAPublicKey):
+        leaf_key_type = "RSA"
+    elif isinstance(leaf_key, dsa.DSAPublicKey):
+        leaf_key_type = "DSA"
+    elif isinstance(leaf_key, ec.EllipticCurvePublicKey):
+        leaf_key_type = "ECDSA"
+    else:
+        leaf_key_type == str(leaf_key.__class__)
 
-        # Beginning and expiration dates of the leaf certificate
-        data['certs']['not_before'] = leaf.not_valid_before
-        data['certs']['not_after'] = leaf.not_valid_after
+    data['certs']['key_type'] = leaf_key_type
 
-        any_sha1_served = False
-        for cert in served_chain:
-            if parse_cert(cert).signature_hash_algorithm.name == "sha1":
-                any_sha1_served = True
+    # Signature of the leaf certificate only.
+    data['certs']['leaf_signature'] = leaf.signature_hash_algorithm.name
 
-        data['certs']['any_sha1_served'] = any_sha1_served
+    # Beginning and expiration dates of the leaf certificate
+    data['certs']['not_before'] = leaf.not_valid_before
+    data['certs']['not_after'] = leaf.not_valid_after
 
-        if data['certs'].get('constructed_issuer'):
-            data['certs']['any_sha1_constructed'] = target['certinfo']['has_sha1_in_certificate_chain']
+    any_sha1_served = False
+    for cert in served_chain:
+        if parse_cert(cert).signature_hash_algorithm.name == "sha1":
+            any_sha1_served = True
 
-    return data
+    data['certs']['any_sha1_served'] = any_sha1_served
+
+    if data['certs'].get('constructed_issuer'):
+        data['certs']['any_sha1_constructed'] = certs.has_sha1_in_certificate_chain
+
+    return data['certs']
 
 
 # Given the cert sub-obj from the sslyze JSON, use
 # the cryptography module to parse its PEM contents.
 def parse_cert(cert):
     backend = cryptography.hazmat.backends.openssl.backend
-    pem_bytes = cert['as_pem'].encode('utf-8')
+    pem_bytes = cert.public_bytes(Encoding.PEM).decode('ascii').encode('utf-8')
```

Comment:
It's possible there's a more efficient way to do this, but I wasn't able to easily figure out one.